proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,520 @@
+
+
+struct Eg10{
+  RWBuffer<int> a;
+  RWBuffer<int> b;
+};
+
+// expected-error@+1{{conflicting register annotations: multiple register annotations detected for register type 'u'}}
+Eg10 e10 : register(u9) : register(u10);
+// expected-error@+1{{conflicting register annotations: multiple register annotations detected for register type 'u'}}
+Eg10 e10a : register(u9, space0) : register(u9, space1);
+
+```
+
+The last common case is if the candidate type is not a valid resource type and not
+a UDT. Types that are or contain a resource are known as "intangible". More generally,
+"intangible types" are "types that have no defined object representation or value 
+representation". In this case, we are dealing with types that cannot be intangible.
+Types that can be immediately determined to not be intangible (that is, types that
+cannot be a resource type or cannot contain a resource type) are types like booleans,
+int, float, float4, etc. If the variable type is among any numerics or a type that
+cannot be an intangible type, there is only one exception where the register keyword
+can legally be applied to the variable. The only valid register type for such a 
+numeric variable type is 'c'. The register keyword in this statement:
+`float f : register(c0)`
+isn't binding a resource, rather it is specifying a constant register binding offset
+within the $Globals cbuffer, which is legacy behavior from DX9. Because this specific 
+register overload is only applicable if the developer wants to make modifications to the
+$Globals buffer, it is not useful when the overload is not used inside the $Globals
+context. That is, if this register overload appears within a `cbuffer {...}` or 
+`tbuffer {...}` scope, the register number will be ignored. In this case, 
+`warn_hlsl_register_type_c_not_in_global_scope` will be emitted, and will be treated as an error by default.
+All other legal register types applied to such a type will emit `err_hlsl_mismatching_register_type_and_variable_type`, 
+except for 'b' or 'i'. If the register type is 'b' or 'i', then `warn_hlsl_deprecated_register_type_b` or 
+`warn_hlsl_deprecated_register_type_i` will be emitted respectively, and treated as errors by default. 
+The compiler will function as if `c` was passed. 
+Invalid register types will cause `err_hlsl_unsupported_register_type_and_variable_type` to be emitted.
+
+Below are some examples:
+
+| Code | Diagnostic |
+|-|-|
+| `float f : register(t0)` | "error: unsupported resource register binding 't' on variable of type 'float'" |
+| `float f : register(c0)` | Valid, no errors
+| `float f : register(b0)` | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding." |
+| `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used." |
+| `float f : register(x0)` | "error: invalid register type 'x' used; expected 't', 'u', 'b', or 's'" |

```

Comment:
Does this need to be updated as part of resolving https://github.com/microsoft/hlsl-specs/pull/230/files#r1681673518?